floydker/src/floydker/render.py

```diff
@@ -0,0 +1,148 @@
+
+    Sample matrix.yml:
+
+    target_1.1.1:
+      _tpl: foo.jinja
+      py2:
+        key1: foo
+        key2: bar
+
+    In the above example, key1 and key2 will be passed in as context to render
+    foo.jinja and generate a `Dockerfile-py2` file under `target_1.1.1`
+    directory.
+    """
+    logger.debug('Entered directory: %s.', matrix_dir)
+    with open(os.path.join(matrix_dir, 'matrix.yml')) as f:
+        try:
+            matrix = yaml.load(f)
+        except:
+            logger.exception('Failed to load matrix from %s', matrix_dir)
+            return
+
+    project_name = os.path.basename(matrix_dir)
+    logger.debug('[%s] Loaded matrix: %s', project_name, matrix)
+
+    for target in matrix.get('$render', []):
+        if not isinstance(target, str):
+            logger.error('target name needs to be a str, %s(%s) found.',
+                         target, type(target))
+            continue
+
+        logger.info('[%s] Loading target <%s>...', project_name, target)
+        # target_cfg here describes how to render multiple images for a given
+        # version of a project
+        target_cfg = matrix[target]
+        tpl_name = target_cfg['_tpl']
+        tpl = jinja2_env.get_template(tpl_name)
+
+        # create target directory for Dockerfiles
+        target_dir = os.path.join(matrix_dir, target)
+        if not os.path.isdir(target_dir):
+            os.mkdir(target_dir)
+
+        for k in target_cfg:
+            if k.startswith('_'):
+                # skip reserved/magic keys
+                continue
+            target_env = k
+            dockerfile_name = 'Dockerfile-' + target_env
+            dockerfile_path = os.path.join(target_dir, dockerfile_name)
+            logger.info('[%s] Rendering target env <%s-%s> to %s...',
+                        project_name, target, target_env, dockerfile_path)
+            target_env_cfg = populate_target_env_cfg(target_cfg, target_env)
+            logger.debug('[%s] Rendering template with context %s',
+                         project_name, target_env_cfg)
+            with open(dockerfile_path, 'w') as f:
+                f.write(tpl.render(**target_env_cfg))
+            logger.debug('[%s] after wrote matrix: %s', project_name, matrix)
+
+
+@click.command()
+@click.argument('search_root')
+@click.option('--project', help='only render project matches the given name')
+@click_log.simple_verbosity_option()
+@click_log.init(__name__)
+def render(search_root, project):
+    tpl_paths = []
+    matrix_dirs = []
+
+    for cur_dir, dirs, files in os.walk(search_root):
+        # TODO: hornor .gitignore
+        for f in files:
+            if f == 'matrix.yml':
+                logger.info('Found matrix in %s', cur_dir)
+                matrix_dirs.append(cur_dir)
+            elif f.endswith('.jinja'):
+                tpl_paths.append(os.path.join(cur_dir, f))
+
+    jinja2_env = jinja2.Environment(loader=FilesLoader(tpl_paths))
+
+    for d in matrix_dirs:
```

Comment:
Can we use full variable names instead of acronyms? `matrix_dir` instead of `d`. Thanks!